### PR TITLE
Cache-related wants (Added --purge)

### PR DESCRIPTION
**Pull request by fabiocarlos97**
Add a new --purge command-line option to remove the cache directory
at $HOME/.cache/cargo-unmaintained/v2. This enables users to easily
clean up the on-disk cache without having to manually find and delete
the directory.

The option is only available on non-Windows platforms with the
on-disk-cache feature enabled, since Windows uses temporary caching.

### CODE REVIEW

**Comment by smoelius** [comment]
@fabiocarlos97 Apologies if you are already working on this, but if you could, please try to get CI to pass. 🙏

**Comment by fabiocarlos97** [comment]
@smoelius hey, I've tried to get all the tests to pass, but can't seem so get dylint test to pass. Also, I'm using Windows so can't really test it out in Linux/ Mac. Could you please advise me on this. Thanks.

**Comment by smoelius** [comment]
The `non_local_effect_before_error_return` lint is triggering because some symbols are made more public by the changes, and that lint checks only publicly accessible symbols by default.

Could you perhaps move the new code into the `on_disk_cache` module? And in lib.rs, do something like this?
```rs
    #[cfg(all(feature = "on-disk-cache", not(windows)))]
    if opts::get().purge {
        return on_disk_cache::purge_cache_directory();
    }
```

I think then no symbol visibilities will have to change and those warnings should go away.

The other lint that's triggering is `inconsistent qualification`. It is objecting to `std::fs::remove_dir_all` because items were imported from `std::fs`, for example. I think if you import `std::fs::remove_dir_all` and use `remove_dir_all` without qualification, that warning should go away.

**Comment by fabiocarlos97** [comment]
@smoelius I've managed to get the CI to pass, please have a look.

File: src/github/mock.rs [hunk 1]
```diff
@@ -18,6 +18,16 @@ impl super::Github for Impl {
             url.as_str()
                 .replace(|c: char| !c.is_ascii_alphanumeric(), "_")
         );
+        // Check if the repository exists
+        let exists_key = format!(
+            "EXISTS_{}",
```
Comment (smoelius) [inline src/github/mock.rs:23]: What is this?

File: src/github/mod.rs [hunk 2]
```diff
@@ -1,8 +1,9 @@
 use super::{RepoStatus, Url};
 use anyhow::Result;
 
-pub(crate) trait Github {
+pub trait Github {
```
Comment (smoelius) [inline src/github/mod.rs:4]: Is it necessary that this symbol's visibility change? Same for the other symbols in this file?

File: src/lib.rs [hunk 3]
```diff
@@ -103,8 +103,7 @@ struct Opts {
     )]
     max_age: u64,
 
-    #[cfg(all(feature = "on-disk-cache", not(windows)))]
-    #[clap(long, help = "Do not cache data on disk for future runs")]
```
Comment (smoelius) [inline src/lib.rs:107]: Was this change to enable local testing on Windows? Can it be rolled back?

File: src/lib.rs [hunk 4]
```diff
@@ -229,16 +235,20 @@ pub fn run() -> Result<()> {
 
     let Cargo {
         subcmd: CargoSubCommand::Unmaintained(opts),
-    } = Cargo::parse_from(args());
+    } = Cargo::parse_from(args().collect::<Vec<_>>());
```
Comment (smoelius) [inline src/lib.rs:238]: Why is this necessary?

File: src/lib.rs [hunk 5]
```diff
@@ -229,16 +235,20 @@ pub fn run() -> Result<()> {
...
 
     opts::init(opts);
 
+    #[cfg(not(windows))]
     if opts::get().save_token {
-        // smoelius: Currently, if additional options are passed besides --save-token, they are
-        // ignored and no error is emitted. This is ugly.
```
Comment (smoelius) [inline src/lib.rs:238]: The changes don't address this comment (do they?). If not, could you please put the comment back in?

File: src/on_disk_cache.rs [hunk 6]
```diff
@@ -54,7 +58,7 @@ struct Entry {
     cloned_url: String,
 }
 
-pub(crate) struct Cache {
+pub struct Cache {
```
Comment (smoelius) [inline src/on_disk_cache.rs:61]: Here also, is it necessary that this symbol's visibility change?

File: src/on_disk_cache.rs [hunk 7]
```diff
@@ -68,25 +72,51 @@ thread_local! {
...
+    };
+    let cache_dir = PathBuf::from(local_app_data)
+        .join("cargo-unmaintained")
+        .join("v2");
+    if let Err(err) = create_dir_all(&cache_dir) {
+        eprintln!("Failed to create cache directory: {err}");
+        exit(1);
+    }
+    cache_dir
+});
```
Comment (smoelius) [inline src/on_disk_cache.rs:107]: It looks like you are trying to support cache directories on Windows(?).

That would be great! But could that please be a separate PR?

File: tests/purge.rs [hunk 8]
```diff
@@ -0,0 +1,42 @@
+#[cfg(all(feature = "on-disk-cache", not(windows)))]
+mod tests {
+    #[test]
+    fn test_purge() {
```
Comment (smoelius) [inline tests/purge.rs:4]: Love this!

**Review by smoelius — COMMENTED** [review]
Thanks a lot for working on this.

One of the first things that jumped out at me was the test, which I think is total rockstar! ⭐

However, I think the changes may be a little more pervasive than they need to be.

I've only reviewed about half of the files. So please forgive me if it takes us a few iterations to get through the review, okay?

File: src/github/mock.rs (hunk 1 above)
Comment (fabiocarlos97) [reply src/github/mock.rs:23]: I added this to test the edge cases in the codebase to handle non-existent repositories, this allows tests to simulate scenarios where repos don't exists as I got an error while running `cargo test --workspace` locally and it gave me a non-existent github repo error, please let me know if I have to strictly remove it.

File: src/github/mod.rs (hunk 2 above)
Comment (fabiocarlos97) [reply src/github/mod.rs:4]: No, it is'nt necessary, I've changed it back to `pub(crate) trait Github`

File: src/lib.rs (hunk 3 above)
Comment (fabiocarlos97) [reply src/lib.rs:107]: Yes, this was done for testing on windows, as you've said to create another PR for adding `--purge` to Windows, I've rolled it back.

File: src/lib.rs (hunk 4 above)
Comment (fabiocarlos97) [reply src/lib.rs:238]: No, it isn't necessary, I've also changed it back.

File: src/lib.rs (hunk 5 above)
Comment (fabiocarlos97) [reply src/lib.rs:238]: I've put the comment back.

File: src/on_disk_cache.rs (hunk 6 above)
Comment (fabiocarlos97) [reply src/on_disk_cache.rs:61]: No, this also wasn't necessary, I've reverted it back

File: src/on_disk_cache.rs (hunk 7 above)
Comment (fabiocarlos97) [reply src/on_disk_cache.rs:107]: You're right. I've adjusted the PR to remove Windows cache directory, I'll create a separate PR for that later

**Comment by fabiocarlos97** [comment]
@smoelius Hey I think I resolved most of your queries, please check to see if any other changes are required, and thank you for being patient with me.

File: src/github/real/util.rs [hunk 9]
```diff
@@ -23,7 +23,7 @@ static CONFIG_DIRECTORY: LazyLock<PathBuf> = LazyLock::new(|| {
 
 static TOKEN_PATH: LazyLock<PathBuf> = LazyLock::new(|| CONFIG_DIRECTORY.join("token.txt"));
 
-pub(super) static PERSONAL_TOKEN: OnceLock<String> = OnceLock::new();
+pub(crate) static PERSONAL_TOKEN: OnceLock<String> = OnceLock::new();
```
Comment (smoelius) [inline src/github/real/util.rs:26]: ^

File: src/github/real/util.rs [hunk 10]
```diff
@@ -62,7 +62,8 @@ pub fn load_token(f: impl FnOnce(&str) -> Result<()>) -> Result<bool> {
     Ok(true)
 }
 
-pub(crate) fn save_token() -> Result<()> {
+#[allow(dead_code)]
+pub fn save_token() -> Result<()> {
```
Comment (smoelius) [inline src/github/real/util.rs:66]: ^

Also, `#[allow(dead_code)]` is unnecessary, maybe?

File: src/lib.rs [hunk 11]
```diff
@@ -33,7 +33,7 @@ pub mod github;
 pub mod packaging;
 
 mod curl;
-mod on_disk_cache;
+pub mod on_disk_cache;
```
Comment (smoelius) [inline src/lib.rs:36]: ^

File: src/lib.rs [hunk 12]
```diff
@@ -233,12 +239,19 @@ pub fn run() -> Result<()> {
 
     opts::init(opts);
 
+    #[cfg(not(windows))]
```
Comment (smoelius) [inline src/lib.rs:242]: Can this be removed?

File: src/on_disk_cache.rs [hunk 13]
```diff
@@ -15,10 +15,6 @@
 //! - A url associated with the package was successfully cloned.
 //! - The clone was performed no more than `refresh_age` days ago.
 //!
-//! A package's entry is considered current if both of the following conditions are met:
-//! - A url associated with the package was successfully cloned.
-//! - The clone was performed no more than `refresh_age` days ago.
-//!
```
Comment (smoelius) [inline src/on_disk_cache.rs:21]: Can this comment be put back in?

File: src/on_disk_cache.rs [hunk 14]
```diff
@@ -40,6 +36,7 @@ use std::{
     sync::LazyLock,
     time::{Duration, SystemTime},
 };
+
```
Comment (smoelius) [inline src/on_disk_cache.rs:39]:
```suggestion
```

File: src/on_disk_cache.rs [hunk 15]
```diff
@@ -68,25 +65,51 @@ thread_local! {
...
+    }
+    cache_dir
+});
+
 #[allow(clippy::unwrap_used)]
 static CRATES_IO_SYNC_CLIENT: LazyLock<SyncClient> =
     LazyLock::new(|| SyncClient::new(USER_AGENT, RATE_LIMIT).unwrap());
 
-pub fn with_cache<T>(f: impl FnOnce(&mut Cache) -> T) -> T {
+pub(crate) fn with_cache<T>(f: impl FnOnce(&mut Cache) -> T) -> T {
```
Comment (smoelius) [inline src/on_disk_cache.rs:106]: ^

File: src/on_disk_cache.rs [hunk 16]
```diff
@@ -68,25 +65,51 @@ thread_local! {
...
     CACHE_ONCE_CELL.with_borrow_mut(|once_cell| {
         let _: &Cache = once_cell.get_or_init(|| {
-            #[cfg(all(feature = "on-disk-cache", not(windows)))]
+            #[cfg(feature = "on-disk-cache")]
```
Comment (smoelius) [inline src/on_disk_cache.rs:109]: Undo for now?

File: src/on_disk_cache.rs [hunk 17]
```diff
@@ -68,25 +65,51 @@ thread_local! {
...
             let temporary = crate::opts::get().no_cache;
 
-            #[cfg(any(not(feature = "on-disk-cache"), windows))]
+            #[cfg(not(feature = "on-disk-cache"))]
```
Comment (smoelius) [inline src/on_disk_cache.rs:112]: Same here?

File: src/url.rs [hunk 18]
```diff
@@ -5,24 +5,24 @@ use std::sync::LazyLock;
 static RE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"^https://[^/]*/[^/]*/[^/]*").unwrap());
 
 #[derive(Clone, Copy, Debug, Default, Eq, Hash, Ord, PartialEq, PartialOrd)]
-pub(crate) struct Url<'a>(&'a str);
+pub struct Url<'a>(&'a str);
```
Comment (smoelius) [inline src/url.rs:8]: Maybe all the changes to this file can be reverted?

File: src/on_disk_cache.rs [hunk 19]
```diff
@@ -366,12 +394,57 @@ impl Cache {
...
+    fn ensure_cache_dir(&self) -> Result<()> {
+        Ok(())
+    }
+}
+
+#[cfg(all(feature = "on-disk-cache", not(windows)))]
+pub(crate) fn purge_cache_directory() {
+    use std::fs::remove_dir_all;
+    if let Err(err) = remove_dir_all(&*CACHE_DIRECTORY) {
+        eprintln!("Failed to remove cache directory: {err}");
```
Comment (smoelius) [inline src/on_disk_cache.rs:447]: Can this be written in the style `remover_dir_all(..).with_context(..)`?

Just using `eprintln` will not cause an error to bubble up and will not set the exit code. I know it's pedantic, but I'd prefer the code be consistent this way.

**Review by smoelius — COMMENTED** [review]
I have another round of comments. Thanks for all of your hard work on this.

File: src/lib.rs (hunk 3 above)
Comment (smoelius) [reply src/lib.rs:107]: Is there a reason to prefer the new phrasing, "Do not use persistent cache"?

**Comment by fabiocarlos97** [comment]
@smoelius I've implemented all of the changes you suggested, including some roll backs and completely re-wrote the purge function to remove any windows implementations.

File: src/on_disk_cache.rs [hunk 20]
```diff
@@ -406,3 +406,35 @@ fn branch_name(repo_dir: &Path) -> Result<String> {
...
+        remove_dir_all(&*CACHE_DIRECTORY).with_context(|| {
+            format!(
+                "failed to remove cache directory at `{}`",
+                CACHE_DIRECTORY.display()
+            )
+        })?;
+
+        println!("Cache directory removed: {}", CACHE_DIRECTORY.display());
+    } else {
+        println!(
```
Comment (smoelius) [inline src/on_disk_cache.rs:433]: Could you please change this `println` and the one on line 431 to `eprintln`, and then squash your commits down to one? Then I think this should be good to merge.

Comment (fabiocarlos97) [reply src/on_disk_cache.rs]: I've implemented the changes.

**Comment by fabiocarlos97** [comment]
@smoelius Hey, I'm not able to merge the PR, could you please squash merge it.

**Comment by smoelius** [comment]
Thanks a lot for this!
